### PR TITLE
Read from MYMETA.{json,yml}

### PR DESCRIPTION
**Pull request by perlpunk**
_Note: I will run this branch for a week or so on the autoupdate host to check if the updates look ok._

Issue: https://progress.opensuse.org/issues/92122

I put the code to read META.{json,yml} into subroutines, so I could reuse
the code for MYMETA.

Along the way I did many small refactorings.

Now intrusive.pl and Intrusive.pm are not needed anymore.

### CODE REVIEW

File: cpanspec [hunk 1]
```diff
@@ -485,18 +484,18 @@ sub get_content(%) {
...
+        my @tmp = split '/', $path; # YAML PP LibYAML
+        my $last = pop @tmp;        # LibYAML
+        push(@pmfiles, "lib/$last.pod", "lib/$last.pm"); # lib/LibYAML.{pm,pod}
     }
     do {
         push(@pmfiles, "$path.pod", "$path.pm");
-    } while ($path =~ s,^[^/]+/,,);
-    push(@pmfiles, "$args{module}")
+    } while ($path =~ s,^[^/]+/,,);   # PP/LibYAML -> LibYAML
+    push(@pmfiles, "$args{module}")   # DateTime
```
[the remark this answers is not in the document]
Comment (coolo) [reply cpanspec:498]: You ask me too much here - I didn't touch this code this decade. Sorry

**Review by perlpunk — COMMENTED** [review]
A problem with a `v` version came up:
https://build.opensuse.org/request/show/960460
```
BuildRequires: perl(Parallel::ForkManager) >= v0.7.6
```
I need to find out why such versions weren't a problem before.

**Comment by perlpunk** [comment]
Actually we do already have modules where we require versions with a leading `v`:
```
% ack "BuildRequires:.* v\d"
perl-Class-DBI-Sweet/perl-Class-DBI-Sweet.spec
33:BuildRequires:  perl(Class::DBI) >= v3.0.12

perl-Config-Model/perl-Config-Model.spec
46:BuildRequires:  perl(Parse::RecDescent) >= v1.90.0

perl-Devel-Cover-Report-Codecov/perl-Devel-Cover-Report-Codecov.spec
50:BuildRequires:  perl(Test::Mock::Time) >= v0.1.6

perl-Module-CPANTS-Analyse/perl-Module-CPANTS-Analyse.spec
42:BuildRequires:  perl(File::Find::Object) >= v0.2.1

perl-Net-Stomp/perl-Net-Stomp.spec
39:BuildRequires:  perl(Test::NiceDump) >= v1.0.0

perl-Number-Phone/perl-Number-Phone.spec
37:BuildRequires:  perl(Parallel::ForkManager) >= v0.7.6

perl-Pod-Tree/perl-Pod-Tree.spec
39:BuildRequires:  perl(Test::Compile) >= v1.2.1

perl-Text-Xslate/perl-Text-Xslate.spec
39:BuildRequires:  perl(Mouse) >= v2.5.0

perl-Text-Table-Manifold/perl-Text-Table-Manifold.spec
31:BuildRequires:  perl(Const::Exporter) >= v0.2.3
```

That is something that cpanspec needs to handle in general, but this PR didn't introduce it, it just revealed the problem. I guess nobody noticed the wrong versions in the other modules so far.

I created the ticket with what we could do:
https://progress.opensuse.org/issues/108120

So the next step for cpanspec is to remove the leading `v` (and this is already done for the package version of the perl distribution itself).

I should do this in a seperate PR.

File: cpanspec [hunk 2]
```diff
@@ -951,127 +959,24 @@ for my $ofile (@args) {
...
 
+    $description = $config->{description} if $config->{description};
     if (!$description) {
-	get_description($content) || get_description($content, 'OVERVIEW');
+        get_description($content) || get_description($content, 'OVERVIEW');
     }
 
-    $summary = get_summary($summary, $content, $module) unless defined $summary;
+    # $author //= get_author($content);
+    # my $authors="Authors:\n--------\n$author";
```
Comment (kalikiana) [inline cpanspec:968]: Did you leave those comments by accident?

File: cpanspec [hunk 3]
```diff
@@ -1895,6 +1681,211 @@ sub decode_latin_or_utf8 {
     return $string;
 }
 
+sub readmetajson {
```
Comment (kalikiana) [inline cpanspec:1684]:
```suggestion
sub read_meta_json {
```
Minor point, but imho this is more readable with underscores

File: cpanspec [hunk 4]
```diff
@@ -1895,6 +1681,211 @@ sub decode_latin_or_utf8 {
...
+    }
+    $stats->{license}->{metajson} = $metajson->{license};
+    if ($metajson->{abstract}) {
+        $stats->{abstract}->{metajson} = $metajson->{abstract};
+    }
+
+    return \%results;
+}
+
+sub readmetayaml {
```
Comment (kalikiana) [inline cpanspec:1730]:
```suggestion
sub read_meta_yaml {
```
Imho more readable with underscores

**Review by kalikiana — CHANGES_REQUESTED** [review]

File: cpanspec [hunk 5]
```diff
@@ -1905,4 +1896,50 @@ sub dump_statistics {
...
+        $sh = new FileHandle ">$script";
+    }
+    else {
+        $sh = new FileHandle $script, O_WRONLY | O_CREAT | O_EXCL;
+    }
+    die "Failed to create $script: $!\n" if (!$sh);
+
+    print $sh "#!/bin/sh\n\n" . "\@\@PERL_PROV\@\@ \"\$\@\" | sed -e '/^$filter_provides[0]\$/d'";
+    if (@filter_provides > 1) {
+        for my $dep (@filter_provides[1 .. $#filter_provides]) {
```
Comment (Martchus) [inline cpanspec:1938]: If `@filter_provides` isn't used anymore one you just `shift`-away the first element and use a simple loop.

**Review by Martchus — APPROVED** [review]
It looks like you know what you're doing. I only have some coding style remarks but I suppose at this point they're not very important.

File: cpanspec [hunk 6]
```diff
@@ -1895,6 +1681,211 @@ sub decode_latin_or_utf8 {
...
+            $license = "SUSE-Public-Domain";    # rpmlint will complain
+        }
+        elsif ($meta->{license} =~ /^restrictive$/i) {
+            $license = "SUSE-NonFree";
+            warn "License is 'restrictive'." . "  This package should not be redistributed.\n";
+        }
+        elsif ($meta->{license} =~ /^unknown$/i) {
+            # do nothing, it's unknown and we know
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply cpanspec:1823]: Yep. But I just moved this code around, and cannot make everything nicer :)

File: cpanspec [hunk 7]
```diff
@@ -1905,4 +1896,50 @@ sub dump_statistics {
...
+}
+
+sub filter_provides {
+    my ($name) = @_;
+    my $script = "$name-filter-provides.sh";
+    verbose "Writing $script...";
+    my $sh;
+    if ($force) {
+        rename($script, "$script~") if (-e $script);
+        $sh = new FileHandle ">$script";
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply cpanspec:1929]: Yep, see above :)

File: cpanspec (hunk 5 above)
Comment (perlpunk) [reply cpanspec:1938]: I don't understand that sentence. But I also have no idea what this code and the --filter-provides option is doing at all and I just moved it into a subroutine (so I can forget about it ;-)

File: cpanspec (hunk 3 above)
Comment (perlpunk) [reply cpanspec]: done

**Review by kalikiana — APPROVED** [review]

File: cpanspec (hunk 5 above)
Comment (kalikiana) [reply cpanspec:1938]: Considering the lack of tests, I second leaving it. Simplifying can backfire if you don't know the code

**Pull request merged**